### PR TITLE
tui: fix problem of using up the TermKey buffer

### LINKED ISSUE #13551: hangs at 100% cpu when holding the escape key too long
- `nvim --version`: NVIM v0.4.4
- `vim -u DEFAULTS` (version: ) behaves differently? yes
- Operating system/version: arch
- Terminal name/version: kitty
- `$TERM`: xterm-kitty

### Steps to reproduce using `nvim -u NORC`

nvim -u NORC
Press and hold the escape key for approximately one minute.

### Actual behaviour

nvim will become unresponsive and the process needs to be kill -9ed.  On my laptop the process will use 100% cpu when it is in this state.

### Expected behaviour

for nothing to happen.

### PR DESCRIPTION
**Pull request by erw7**
fixes https://github.com/neovim/neovim/issues/13551#issuecomment-747720618.

- [X] Add test.

### CODE REVIEW

File: test/functional/terminal/tui_spec.lua [hunk 1]
```diff
@@ -875,6 +875,11 @@ describe('TUI', function()
       {3:-- TERMINAL --}                                    |
     ]=])
   end)
+
+  it("does not crash on successive <ESC> inputs #13551", function()
+    command([[call chansend(b:terminal_job_id, repeat("\<ESC>", 257))]])
+    screen:expect_unchanged()
```
Comment (vigoux) [inline test/functional/terminal/tui_spec.lua:881]: I think that the recommendation is using `alive()` here?

Comment (erw7) [reply test/functional/terminal/tui_spec.lua:881]: I think what we need to check is whether nvim in the terminal is alive and accepting input or not, so I did the following.

```lua
command([[call chansend(b:terminal_job_id, ":echo 'alive'\<CR>")]])
screen:expect{any='alive'}
```

**Comment by erw7** [comment]
It seems that this fix is still not enough.

```vim
nvim -u NORC
:term nvim -u NORC
:call chansend(b:terminal_job_id, repeat("\<ESC>", 255).."\<ESC>OP")
" Insert mode is entered and P is input.
```

```vim
nvim -u NORC
:term nvim -u NORC
:call chansend(b:terminal_job_id, repeat("\<ESC>", 254).."\<ESC>OP")
" The ASAN build terminates abnormally with the following log.
```

<details>
<summary>ASAN log</summary>
<pre><code>
==1088==ERROR: AddressSanitizer: heap-use-after-free on address 0x611000010140 at pc 0x00000046c20d bp 0x7fa5fe20b240 sp 0x7fa5fe20aa00
READ of size 256 at 0x611000010140 thread T1
==1088==ERROR: AddressSanitizer: heap-use-after-free on address 0x611000010140 at pc 0x00000046c20d bp 0x7fa5fe20b240 sp 0x7fa5fe20aa00
READ of size 256 at 0x611000010140 thread T1
    #0 0x46c20c in memmove (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x46c20c)
    #0 0x46c20c in memmove (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x46c20c)
    #1 0x7fa60213073d in termkey_push_bytes (/usr/local/lib/libtermkey.so.1+0x573d)
    #2 0x1a5a34b in handle_raw_buffer /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/input.c:605:25
    #3 0x1a5777d in tinput_read_cb /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/input.c:639:3
    #4 0xb894f0 in read_event /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/rstream.c:194:5
    #5 0xb8853f in invoke_read_cb /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/rstream.c:207:3
    #6 0xb86ee8 in read_cb /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/rstream.c:135:3
    #1 0x7fa60213073d in termkey_push_bytes (/usr/local/lib/libtermkey.so.1+0x573d)
    #2 0x1a5a34b in handle_raw_buffer /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/input.c:605:25
    #3 0x1a5777d in tinput_read_cb /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/input.c:639:3
    #4 0xb894f0 in read_event /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/rstream.c:194:5
    #5 0xb8853f in invoke_read_cb /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/rstream.c:207:3
    #6 0xb86ee8 in read_cb /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/rstream.c:135:3
    #7 0x7fa6021e2fb0 in uv__read /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/stream.c:1239:7
    #7 0x7fa6021e2fb0 in uv__read /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/stream.c:1239:7
    #8 0x7fa6021e35c7 in uv__stream_io /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/stream.c:1306:5
    #8 0x7fa6021e35c7 in uv__stream_io /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/stream.c:1306:5
    #9 0x7fa6021e8e7b in uv__io_poll /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/linux-core.c:462:11
    #9 0x7fa6021e8e7b in uv__io_poll /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/linux-core.c:462:11
    #10 0x7fa6021d90cb in uv_run /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/core.c:385:5
    #10 0x7fa6021d90cb in uv_run /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/core.c:385:5
    #11 0xb6fcbb in loop_poll_events /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/loop.c:62:3
    #11 0xb6fcbb in loop_poll_events /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/loop.c:62:3
    #12 0x1a7a89e in tui_main /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/tui.c:455:5
    #12 0x1a7a89e in tui_main /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/tui.c:455:5
    #13 0x1ae3c9d in ui_thread_run /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/ui_bridge.c:105:3
    #13 0x1ae3c9d in ui_thread_run /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/ui_bridge.c:105:3
    #14 0x7fa6021a2608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x9608)
    #14 0x7fa6021a2608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x9608)
    #15 0x7fa601e02292 in clone (/lib/x86_64-linux-gnu/libc.so.6+0x122292)
    #15 0x7fa601e02292 in clone (/lib/x86_64-linux-gnu/libc.so.6+0x122292)


0x611000010160 is located 0 bytes to the right of 224-byte region [0x611000010080,0x611000010160)
0x611000010160 is located 0 bytes to the right of 224-byte region [0x611000010080,0x611000010160)
freed by thread T1 here:
freed by thread T1 here:
    #0 0x4ce329 in realloc (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x4ce329)
    #0 0x4ce329 in realloc (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x4ce329)
    #1 0x115c722 in xrealloc /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/memory.c:155:15
    #1 0x115c722 in xrealloc /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/memory.c:155:15
    #2 0x1a70d91 in tui_hl_attr_define /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/tui.c:1190:3
    #2 0x1a70d91 in tui_hl_attr_define /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/tui.c:1190:3
    #3 0x1ae9f13 in ui_bridge_hl_attr_define_event /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/ui_bridge.c:150:3
    #3 0x1ae9f13 in ui_bridge_hl_attr_define_event /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/ui_bridge.c:150:3
    #4 0xb76c4f in multiqueue_process_events /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/multiqueue.c:157:7
    #4 0xb76c4f in multiqueue_process_events /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/multiqueue.c:157:7
    #5 0xb700d1 in loop_poll_events /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/loop.c:70:3
    #5 0xb700d1 in loop_poll_events /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/event/loop.c:70:3
    #6 0x1a7a7fe in tui_main /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/tui.c:448:12
    #6 0x1a7a7fe in tui_main /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/tui.c:448:12
    #7 0x1ae3c9d in ui_thread_run /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/ui_bridge.c:105:3
    #7 0x1ae3c9d in ui_thread_run /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/ui_bridge.c:105:3
    #8 0x7fa6021a2608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x9608)
    #8 0x7fa6021a2608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x9608)


previously allocated by thread T1 here:
previously allocated by thread T1 here:
    #0 0x4ce329 in realloc (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x4ce329)
    #0 0x4ce329 in realloc (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x4ce329)
    #1 0x115c722 in xrealloc /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/memory.c:155:15
    #1 0x115c722 in xrealloc /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/memory.c:155:15
    #2 0x1a79b81 in tui_main /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/tui.c:432:3
    #2 0x1a79b81 in tui_main /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/tui/tui.c:432:3
    #3 0x1ae3c9d in ui_thread_run /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/ui_bridge.c:105:3
    #3 0x1ae3c9d in ui_thread_run /home/erw7/src/github.com/neovim/neovim/build/../src/nvim/ui_bridge.c:105:3
    #4 0x7fa6021a2608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x9608)
    #4 0x7fa6021a2608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x9608)


Thread T1 created by T0 here:
Thread T1 created by T0 here:
    #0 0x4b8dba in pthread_create (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x4b8dba)
    #0 0x4b8dba in pthread_create (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x4b8dba)
    #1 0x7fa6021e4ef1 in uv_thread_create_ex /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/thread.c:259:9
    #1 0x7fa6021e4ef1 in uv_thread_create_ex /home/erw7/src/github.com/libuv/libuv/build-wsl-Ubuntu-20.04/../src/unix/thread.c:259:9


SUMMARY: AddressSanitizer: heap-use-after-free (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x46c20c) in memmove
SUMMARY: AddressSanitizer: heap-use-after-free (/home/erw7/src/github.com/neovim/neovim/build/bin/nvim+0x46c20c) in memmove
Shadow bytes around the buggy address:
  0x0c227fff9fd0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c227fff9fe0: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c227fff9ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffa000: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x0c227fffa010: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c227fffa020: fd fd fd fd fd fd fd fd[fd]fd fd fd fa fa fa fa
  0x0c227fffa030: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c227fffa040: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffa050: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x0c227fffa060: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffa070: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
Shadow bytes around the buggy address:
  0x0c227fff9fd0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c227fff9fe0: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c227fff9ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffa000: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x0c227fffa010: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c227fffa020: fd fd fd fd fd fd fd fd[fd]fd fd fd fa fa fa fa
  0x0c227fffa030: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c227fffa040: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffa050: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x0c227fffa060: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffa070: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==1088==ABORTING
==1088==ABORTING
</code></pre>
</details>

I think it is a bug in libtermkey that `termkey_getkey()` returns `TERMKEY_RES_AGAIN` for `<ESC><ESC><ESC>` input.

CC: @leonerd.

**Comment by erw7** [comment]
When `termkey_push_bytes()` returns -1, I thought it would be a good idea to call `termkey_getkey_force()` until there is enough space in the buffer, but I found a new bug in libtermkey.

```vim
nvim -u NORC
:term nvim -u NORC
:call chansend(b:terminal_job_id, ":let g:esc = 0 | nnoremap <silent><ESC> :let g:esc += 1<CR>\<CR>")
:call chansend(b:terminal_job_id, repeat("\<ESC>", 50) .. ":echo g:esc\<CR>")
1 " This should be 50.
```

`termkey_getkey_force()` will only generate one `<ESC>` from `<ESC><ESC><ESC>...`. It seems that the only way to fix this problem completely is to fix libtermkey.

Since the actual input probably won't generate 255 <ESC> followed by <F1> and so on, we may as well merge this as a workaround until libtermkey is fixed.

**Comment by leonerd** [comment]
Oops. I'll take a look

**Pull request closed without merging**